src/software/simulation/threaded_simulator.h

```diff
@@ -0,0 +1,99 @@
+     *
+     * @param field The field to initialize the simulation with
+     */
+    explicit ThreadedSimulator(const Field& field);
+    ~ThreadedSimulator();
+
+    /**
+     * Registers the given callback function. This callback function will be
+     * called each time the simulation updates and a new SSL_WrapperPacket
+     * is generated.
+     *
+     * @param callback The callback function to register
+     */
+    void registerOnSSLWrapperPacketReadyCallback(const std::function<void(SSL_WrapperPacket)>& callback);
+
+    /**
+     * Starts running the simulator in a new thread. This is a non-blocking call.
+     * If the simulator is already running, this function does nothing.
+     */
+    void startSimulation();
+
+    /**
+     * Stops the simulation if it is running. If the simulation is not already
+     * running, this function does nothing.
+     */
+    void stopSimulation();
+
+    /**
+     * Sets the state of the ball in the simulation. No more than 1 ball may exist
+     * in the simulation at a time. If a ball does not already exist, a ball
+     * is added with the given state. If a ball already exists, it's state is set to the
+     * given state.
+     *
+     * @param ball_state The new ball state
+     */
+    void setBallState(const BallState& ball_state);
+
+    /**
+     * Removes the ball from the physics world. If a ball does not already exist,
+     * this has no effect.
+     */
+    void removeBall();
+
+    /**
+     * Adds robots to the specified team with the given initial states.
+     *
+     * @pre The robot IDs must not be duplicated and must not match the ID
+     * of any robot already on the specified team.
+     *
+     * @throws runtime_error if any of the given robot ids are duplicated, or a
+     * robot already exists on the specified team with one of the new IDs
+     *
+     * @param robots the robots to add
+     */
+    void addYellowRobots(const std::vector<RobotStateWithId>& robots);
+    void addBlueRobots(const std::vector<RobotStateWithId>& robots);
+
+    /**
+     * Sets the primitives being simulated by the robots in simulation
+     *
+     * @param primitives The primitives to simulate
+     */
+    void setYellowRobotPrimitives(ConstPrimitiveVectorPtr primitives);
+    void setBlueRobotPrimitives(ConstPrimitiveVectorPtr primitives);
+
+private:
+    /**
+     * The function that runs inside the simulation thread, handling
+     * updating the simulation and calling the callback functions
+     */
+    void runSimulationLoop();
+
+    std::vector<std::function<void(SSL_WrapperPacket)>> ssl_wrapper_packet_callbacks;
+
+    Simulator simulator;
+    std::mutex simulator_mutex;
+    std::thread simulation_thread;
+    std::atomic_bool simulation_thread_started;
+    std::atomic_bool stopping_simulation;
+
+    // 60HZ is approximately the framerate of the real-life cameras
+    static constexpr double TIME_STEP_SECONDS = 1.0 / 60.0;
```

Comment:
Isn't this suplicated with the other simulator PR? Can we put it in a common constants file?

Reply:
Sort of but not really. We may want to run them at different speeds since they're pretty different use-cases so each "user" of the `Simulator` object can define its own timing. That's why I left them separate.